discord-scripts/invite-management.ts

```diff
@@ -0,0 +1,181 @@
+import { Robot } from "hubot"
+import {
+  Client,
+  TextChannel,
+  VoiceChannel,
+  StageChannel,
+  ApplicationCommandOptionType,
+} from "discord.js"
+
+async function createInvite(
+  discordClient: Client,
+  channelId: string,
+  maxAge: number | string = 86400,
+  maxUses: number | string = 0,
+): Promise<string | null> {
+  if (!discordClient) {
+    throw new Error("Discord client is not initialized.")
+  }
+
+  let channelOrigin: TextChannel | VoiceChannel | StageChannel | null = null
+  try {
+    const fetchedChannel = await discordClient.channels.fetch(channelId)
+    if (
+      fetchedChannel instanceof TextChannel ||
+      fetchedChannel instanceof VoiceChannel ||
+      fetchedChannel instanceof StageChannel
+    ) {
+      channelOrigin = fetchedChannel
+    }
+  } catch (error) {
+    throw new Error("Channel not found or access denied.")
+  }
+
+  if (!channelOrigin) {
+    return null
+  }
```

Comment:
TypeScript's control flow analysis should be able to recognize that this version will work the same way:

```suggestion
  const channelOrigin = await discordClient.channels.fetch(channelId)
  
  if (
    channelOrigin === null ||
    (!(channelOrigin instanceof TextChannel) &&
    !(channelOrigin instanceof VoiceChannel) &&
    !(channelOrigin instanceof StageChannel))
  ) {
   throw new Error("Channel not found or access denied.")
  }
```

This is because TypeScript will see that the above control flow will throw if the fetched channel is null or not one of the 3 types, so if control flows past this statement the channel *must* be one of those three types.

We can actually get even fancier here if we want to though… We can filter only for channels that allow invites:

```suggestion
  const channelOrigin = await discordClient.channels.fetch(channelId)
  
  if (
    channelOrigin === null ||
    !("createInvite" in channelOrigin)
  ) {
   throw new Error("Channel not found or access denied.")
  }
```

I might also consider renaming `channelOrigin` for clarity—perhaps to `channelForInvite` or so?

Reply:
Thanks for the feedback on this and it's a good way to stop my old JS habits, the simplicity of the last commit suggestion is 💯 🤩